File: src/main/java/org/example/database/DatabaseProbe.java
package org.example.database;

import org.example.model.DbResult;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.time.LocalDateTime;

public class DatabaseProbe {

    private final String dbUrl;
    private final String username;
    private final String password;

    public DatabaseProbe(String dbUrl,
                         String username,
                         String password) {
        this.dbUrl = dbUrl;
        this.username = username;
        this.password = password;
    }

    public DbResult executeProbe() {

        LocalDateTime timestamp = LocalDateTime.now();

        double dbConnectTimeMs = 0.0;
        double queryTimeMs = 0.0;

        boolean success = false;
        String errorMessage = null;

        try {

            long connectionStartTime = System.nanoTime();

            Connection connection = DriverManager.getConnection(
                    dbUrl,
                    username,
                    password
            );

            long connectionEndTime = System.nanoTime();

            dbConnectTimeMs =
                    (connectionEndTime - connectionStartTime)
                            / 1_000_000.0;

            long queryStartTime = System.nanoTime();

            try (PreparedStatement preparedStatement =
                         connection.prepareStatement("SELECT 1");

                 ResultSet resultSet =
                         preparedStatement.executeQuery()) {

                if (resultSet.next()) {
                    resultSet.getInt(1);
                }
            }

            long queryEndTime = System.nanoTime();

            queryTimeMs =
                    (queryEndTime - queryStartTime)
                            / 1_000_000.0;

            success = true;

            connection.close();

        } catch (Exception e) {

            success = false;
            errorMessage = e.getMessage();
        }

        return new DbResult(
                timestamp,
                dbConnectTimeMs,
                queryTimeMs,
                success,
                errorMessage
        );
    }
}

File: node-version/src/probes/DatabaseProbe.js
import { Client } from 'pg';

import DbResult from '../model/DbResult.js';

class DatabaseProbe {

    constructor(
        host,
        port,
        database,
        user,
        password
    ) {
        this.host = host;
        this.port = port;
        this.database = database;
        this.user = user;
        this.password = password;
    }

    async executeProbe() {

        const timestamp = new Date();

        let dbConnectTimeMs = 0;
        let queryTimeMs = 0;

        let success = false;
        let errorMessage = null;

        let client;

        try {

            client = new Client({
                host: this.host,
                port: this.port,
                database: this.database,
                user: this.user,
                password: this.password
            });

            const connectionStart =
                process.hrtime.bigint();

            await client.connect();

            const connectionEnd =
                process.hrtime.bigint();

            dbConnectTimeMs =
                Number(connectionEnd - connectionStart)
                / 1_000_000;

            const queryStart =
                process.hrtime.bigint();

            await client.query('SELECT 1');

            const queryEnd =
                process.hrtime.bigint();

            queryTimeMs =
                Number(queryEnd - queryStart)
                / 1_000_000;

            success = true;

        } catch (error) {

            success = false;
            errorMessage = error.message;

        } finally {

            if (client) {
                try {
                    await client.end();
                } catch (_) {
                }
            }
        }

        return new DbResult(
            timestamp,
            dbConnectTimeMs,
            queryTimeMs,
            success,
            errorMessage
        );
    }
}

export default DatabaseProbe;